org-code-javabuilder/lib/src/main/java/dev/javabuilder/LocalContentManager.java

```diff
@@ -0,0 +1,94 @@
+package dev.javabuilder;
+
+import static dev.javabuilder.LocalWebserverConstants.DIRECTORY;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.code.javabuilder.*;
+import org.code.protocol.ContentManager;
+import org.code.protocol.InternalErrorKey;
+import org.code.protocol.JavabuilderException;
+import org.code.protocol.LoggerUtils;
+import org.json.JSONException;
+
+public class LocalContentManager implements ContentManager, ProjectFileLoader {
+  private static final String SERVER_URL_FORMAT = "http://localhost:8080/%s/%s";
+
+  private ProjectData projectData;
+
+  public LocalContentManager() {
+    this.projectData = null;
+  }
+
+  // Used for testing
+  LocalContentManager(ProjectData projectData) {
+    this.projectData = projectData;
+  }
+
+  @Override
+  public UserProjectFiles loadFiles() throws InternalServerError, UserInitiatedException {
+    this.loadProjectDataIfNeeded();
+    return this.projectData.getSources();
+  }
+
+  @Override
+  public String generateAssetUrl(String filename) {
+    try {
+      this.loadProjectDataIfNeeded();
+    } catch (InternalServerError e) {
+      // Log this error since this indicates an issue reading project data
+      LoggerUtils.logException(e);
+      return null;
+    }
+    return this.projectData.getAssetUrl(filename);
+  }
+
+  @Override
+  public String generateAssetUploadUrl(String filename) throws InternalServerError {
+    this.loadProjectDataIfNeeded();
+    final String uploadUrl = String.format(SERVER_URL_FORMAT, DIRECTORY, filename);
+    // Add new asset URL to our project data so it can be referenced later.
```

Comment:
I got a bit confused by us putting the upload url in the list of urls we use for get. Looks like in this case it's the same url, mind adding a comment explaining that?

Reply:
Yep, I'll add some more context in the comments. Since this is all on the local fileserver, the PUT url becomes the same as the GET url once the object is uploaded.